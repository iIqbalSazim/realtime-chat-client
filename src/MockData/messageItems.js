export const mockMessages = [
  { text: "Hello", senderId: 1 },
  { text: "How are you?", senderId: 1 },
];
