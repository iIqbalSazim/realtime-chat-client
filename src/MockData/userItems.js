export const mockUsers = [
  { id: 1, username: "Test User1" },
  { id: 2, username: "Test User2" },
];
