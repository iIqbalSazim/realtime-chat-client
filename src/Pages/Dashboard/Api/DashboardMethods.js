import { getRequest, postRequest } from "../../../Config/Axios/AxiosConfig";

export const fetchAllUsers = async () => {
  return await getRequest("users");
};

export const fetchUserMessageHistory = async (receiverId) => {
  return await getRequest(`messages?chatting_with=${receiverId}`);
};

export const sendMessage = async (message) => {
  return await postRequest("messages", message);
};
