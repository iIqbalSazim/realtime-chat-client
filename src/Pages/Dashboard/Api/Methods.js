import { getRequest } from "../../../Config/Axios/AxiosConfig";

export const fetchAllUsers = async () => {
  return await getRequest("users");
};

export const fetchUserMessageHistory = async (senderId, receiverId) => {
  return await getRequest(
    `messages?sender_id=${senderId}&receiver_id=${receiverId}`
  );
};
