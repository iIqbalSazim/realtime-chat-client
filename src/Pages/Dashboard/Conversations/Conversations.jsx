import { Conversation, ConversationList } from "@chatscope/chat-ui-kit-react";
import { useDispatch } from "react-redux";

import { changeChatUser } from "../../../Actions/chat";

const Conversations = ({ users }) => {
  const dispatch = useDispatch();

  const handleClick = (user) => {
    dispatch(changeChatUser(user));
  };

  return (
    <>
      {users.length !== 0 ? (
        <ConversationList>
          {users.map((user) => (
            <Conversation
              name={user.username}
              key={user.id}
              onClick={() => handleClick(user)}
            />
          ))}
        </ConversationList>
      ) : (
        <div>No users found</div>
      )}
    </>
  );
};

export default Conversations;
