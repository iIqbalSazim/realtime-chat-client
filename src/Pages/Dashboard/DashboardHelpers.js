export const convertApiMessagesToChatMessages = (
  apiMessages,
  currentUser,
  chattingWith
) => {
  return apiMessages.map((message) => {
    if (message.sender_id === currentUser.id) {
      return {
        id: message.id,
        message: message.content,
        sender: currentUser.username,
        receiver: chattingWith.username,
        sender_id: currentUser.id,
        receiver_id: chattingWith.id,
        direction: "outgoing",
        position: "single",
      };
    } else {
      return {
        id: message.id,
        message: message.content,
        sender: chattingWith.username,
        receiver: currentUser.username,
        sender_id: chattingWith.id,
        receiver_id: currentUser.id,
        direction: "incoming",
        position: "single",
      };
    }
  });
};

export const convertApiMessageToChatMessage = (
  message,
  currentUser,
  chattingWith
) => {
  if (message.sender_id === currentUser.id) {
    return {
      id: message.id,
      message: message.content,
      sender: currentUser.username,
      receiver: chattingWith.username,
      sender_id: currentUser.id,
      receiver_id: chattingWith.id,
      direction: "outgoing",
      position: "single",
    };
  } else {
    return {
      id: message.id,
      message: message.content,
      sender: chattingWith.username,
      receiver: currentUser.username,
      sender_id: chattingWith.id,
      receiver_id: currentUser.id,
      direction: "incoming",
      position: "single",
    };
  }
};
