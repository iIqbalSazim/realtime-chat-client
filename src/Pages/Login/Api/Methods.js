import { postRequest } from "../../../Config/Axios/AxiosConfig";

export const loginUser = async (user) => {
  return await postRequest("login", user);
};

export const revokeToken = async (token) => {
  return await postRequest("revoke_token", token);
};
