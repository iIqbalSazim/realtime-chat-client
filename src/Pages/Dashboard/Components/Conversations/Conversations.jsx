import { useState } from "react";
import { useDispatch } from "react-redux";
import { Conversation, ConversationList } from "@chatscope/chat-ui-kit-react";

import { changeChatUser, setMessages } from "../../../../Stores/Actions/chat";
import { fetchUserMessageHistory } from "../../Api/DashboardMethods";
import { convertApiMessagesToChatMessages } from "../../DashboardHelpers";

const Conversations = ({ currentUser, users }) => {
  const dispatch = useDispatch();

  const [activeChat, setActiveChat] = useState(null);

  const handleClick = async (user) => {
    await getMessageHistory(user);
    setActiveChat(user.id);
    dispatch(changeChatUser(user));
  };

  const getMessageHistory = async (user) => {
    const response = await fetchUserMessageHistory(user.id);

    const apiMessages = response.data;

    const chatMessages = convertApiMessagesToChatMessages(
      apiMessages,
      currentUser,
      user
    );

    dispatch(setMessages([...chatMessages]));
  };

  return (
    <>
      {users.length !== 0 ? (
        <ConversationList>
          {users.map((user) => (
            <Conversation
              name={user.username}
              key={user.id}
              onClick={() => handleClick(user)}
              className={`${activeChat === user.id ? "bg-slate-200" : ""} my-2`}
            />
          ))}
        </ConversationList>
      ) : (
        <div>No users found</div>
      )}
    </>
  );
};

export default Conversations;
