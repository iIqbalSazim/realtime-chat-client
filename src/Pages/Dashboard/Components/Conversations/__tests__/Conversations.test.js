import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { renderWithProviders } from "../../../../../../jest.setup";

import { mockUsers } from "../../../../../MockData/userItems";
import { mockMessages } from "../../../../../MockData/messageItems";
import * as DashboardMethods from "../../../Api/DashboardMethods";
import * as DashboardHelpers from "../../../DashboardHelpers";
import Conversations from "../Conversations";

describe("Conversations", () => {
  test("renders no users found when users array is empty", () => {
    const { container } = renderWithProviders(
      <Conversations currentUser={{}} users={[]} />
    );

    const noUsersFoundText = screen.getByText("No users found");
    expect(noUsersFoundText).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test("renders conversation list with users", async () => {
    const { container } = renderWithProviders(
      <Conversations currentUser={{}} users={mockUsers} />
    );

    const conversationList = screen.getByRole("list");
    const conversationItems = screen.getAllByRole("listitem");

    expect(conversationList).toBeInTheDocument();
    expect(conversationItems.length).toBe(mockUsers.length);

    expect(container).toMatchSnapshot();
  });

  test("handles click on conversation item", async () => {
    const fetchUserMessageHistorySpy = jest.spyOn(
      DashboardMethods,
      "fetchUserMessageHistory"
    );
    const convertApiMessagesToChatMessagesSpy = jest.spyOn(
      DashboardHelpers,
      "convertApiMessagesToChatMessages"
    );

    fetchUserMessageHistorySpy.mockResolvedValueOnce({
      data: mockMessages,
    });
    convertApiMessagesToChatMessagesSpy.mockReturnValueOnce(mockMessages);

    const { container } = renderWithProviders(
      <Conversations currentUser={{}} users={mockUsers} />
    );

    const conversationItem = screen.getByText(/Test User1/i);
    await userEvent.click(conversationItem);

    expect(fetchUserMessageHistorySpy).toHaveBeenCalledWith(mockUsers[0].id);
    expect(convertApiMessagesToChatMessagesSpy).toHaveBeenCalledWith(
      mockMessages,
      {},
      mockUsers[0]
    );

    fetchUserMessageHistorySpy.mockRestore();
    convertApiMessagesToChatMessagesSpy.mockRestore();

    expect(container).toMatchSnapshot();
  });
});
