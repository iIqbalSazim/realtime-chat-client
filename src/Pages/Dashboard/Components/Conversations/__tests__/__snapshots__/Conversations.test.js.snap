// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conversations handles click on conversation item 1`] = `
<div>
  <div
    class="cs-conversation-list"
  >
    <div
      class="scrollbar-container ps"
    >
      <ul>
        <li>
          <div
            class="cs-conversation bg-slate-200 my-2"
          >
            <div
              class="cs-conversation__content"
            >
              <div
                class="cs-conversation__name"
              >
                Test User1
              </div>
              <div
                class="cs-conversation__info"
              />
            </div>
          </div>
        </li>
        <li>
          <div
            class="cs-conversation  my-2"
          >
            <div
              class="cs-conversation__content"
            >
              <div
                class="cs-conversation__name"
              >
                Test User2
              </div>
              <div
                class="cs-conversation__info"
              />
            </div>
          </div>
        </li>
      </ul>
      <div
        class="ps__rail-x"
        style="left: 0px; top: 0px;"
      >
        <div
          class="ps__thumb-x"
          style="left: 0px; width: 0px;"
          tabindex="0"
        />
      </div>
      <div
        class="ps__rail-y"
        style="top: 0px; left: 0px;"
      >
        <div
          class="ps__thumb-y"
          style="top: 0px; height: 0px;"
          tabindex="0"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Conversations renders conversation list with users 1`] = `
<div>
  <div
    class="cs-conversation-list"
  >
    <div
      class="scrollbar-container ps"
    >
      <ul>
        <li>
          <div
            class="cs-conversation  my-2"
          >
            <div
              class="cs-conversation__content"
            >
              <div
                class="cs-conversation__name"
              >
                Test User1
              </div>
              <div
                class="cs-conversation__info"
              />
            </div>
          </div>
        </li>
        <li>
          <div
            class="cs-conversation  my-2"
          >
            <div
              class="cs-conversation__content"
            >
              <div
                class="cs-conversation__name"
              >
                Test User2
              </div>
              <div
                class="cs-conversation__info"
              />
            </div>
          </div>
        </li>
      </ul>
      <div
        class="ps__rail-x"
        style="left: 0px; top: 0px;"
      >
        <div
          class="ps__thumb-x"
          style="left: 0px; width: 0px;"
          tabindex="0"
        />
      </div>
      <div
        class="ps__rail-y"
        style="top: 0px; left: 0px;"
      >
        <div
          class="ps__thumb-y"
          style="top: 0px; height: 0px;"
          tabindex="0"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Conversations renders no users found when users array is empty 1`] = `
<div>
  <div>
    No users found
  </div>
</div>
`;
